### PR TITLE
Add communication service

### LINKED ISSUE #19: Two users in a collab session should be able to chat

### LINKED ISSUE #53: N1: Communication with matched users
* Messaging
 * Call: https://hub.docker.com/r/flashphoner/webcallserver

### PR DESCRIPTION
**Pull request by kflim**
- [x] Add backend for communication service
- [x] Add basic frontend for chatting
- [x] Fix state update for who is typing
- [x] Add proper styling to the chat
- [x] Refactor to use hook

(optional) Add video call
(optional) Move to microservice

https://github.com/CS3219-AY2324S1/ay2324s1-course-assessment-g39/assets/61484904/ab30270f-62c4-4ac7-9979-519096d6f104

### CODE REVIEW

**Comment by Gabau** [comment]
Changed base to view diff

File: src/server/api/routers/communication.ts [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+        sessionId,
+        userId: senderId,
+        userName: senderName,
+        isTyping: false,
+      });
+
+      return messageObject;
+    }),
+
+  subscribeToSessionMessages: protectedProcedure
```
Comment (Gabau) [inline src/server/api/routers/communication.ts:81]: Wouldn't it be better for subscription if we simplify to only one subscription for the communications route.
So when someone is typing we send a typing message, and when someone has sent a message, we send a message object, but we only use one subscription to do that

Comment (Gabau) [reply src/server/api/routers/communication.ts:81]: Basicaly shift the ee.on("typing", ...) code here

File: src/pages/collab/rooms/[id].tsx [hunk 2]
```diff
@@ -207,6 +211,61 @@ const Output = ({
   </div>
 );
 
+const Chatbox = ({
```
Comment (Gabau) [inline src/pages/collab/rooms/[id].tsx]: I think its best if the communication related components be stored in the components folder?

**Review by Gabau — COMMENTED** [review]
Some comments, haven't had the time to run the code tho

**Comment by Gabau** [comment]
Can put a gif/ image of the chat box at work in the pr?

**Comment by kflim** [comment]
> Can put a gif/ image of the chat box at work in the pr?

Ok, I just put a video up showcasing it, dk how to convert to gif with good resolution

File: src/pages/collab/rooms/[id].tsx (hunk 2 above)
Comment (kflim) [reply src/pages/collab/rooms/[id].tsx]: done

Comment (kflim) [reply src/pages/collab/rooms/[id].tsx]: done

File: src/server/api/routers/communication.ts (hunk 1 above)
Comment (kflim) [reply src/server/api/routers/communication.ts:81]: you mean one subscription emits two datatypes? or how about one data type which includes message as well as if the partneristyping

Comment (Gabau) [reply src/server/api/routers/communication.ts:81]: One data type which includes an optional message and partner is typing

File: src/components/ChatBox.tsx [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    userTypingMessage,
+  ] = useSessionComm(sessionId, userId, userName);
+
+  return (
+    <div className={className}>
+      <div className="messages-container overflow-y-auto flex flex-col h-[18.5rem]">
+        {allSessionMessages?.map((message, index) => {
+          return (
+            <div
+              key={index}
```
Comment (Gabau) [inline src/components/ChatBox.tsx]: iirc it's not really recommended to use index as a key, should use some kind of unique id

Comment (Gabau) [reply src/components/ChatBox.tsx]: Also I think its a minor change to flip the sides of the message by checking the message's user id against the current sessions user id

File: prisma/postgres/schema.prisma [hunk 4]
```diff
@@ -174,4 +174,16 @@ model QuestionAttempt {
     user        User         @relation(fields: [userId], references: [id], onDelete: Cascade)
 
     @@unique([userId, questionId, answerId])
+
+}
+
+model SessionMessage {
+    sessionId   String
+    senderId    String
```
Comment (Gabau) [inline prisma/postgres/schema.prisma]: Shouldn't we bind this to user via a relations

**Review by Gabau — COMMENTED** [review]
Some more comments

File: src/components/ChatBox.tsx [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    allSessionMessages,
+    sendMessage,
+    onTyping,
+    currentMessage,
+    userTypingMessage,
+  ] = useSessionComm(sessionId, userId, userName);
+
+  return (
+    <div className={className}>
+      <div className="messages-container overflow-y-auto flex flex-col h-[18.5rem]">
```
Comment (Gabau) [inline src/components/ChatBox.tsx]: idt we should be using fixed height for this div?

File: src/server/api/routers/communication.ts (hunk 1 above)
Comment (kflim) [reply src/server/api/routers/communication.ts:81]: done

File: src/components/ChatBox.tsx (hunk 3 above)
Comment (kflim) [reply src/components/ChatBox.tsx]: both done

File: prisma/postgres/schema.prisma (hunk 4 above)
Comment (kflim) [reply prisma/postgres/schema.prisma]: done

File: src/components/ChatBox.tsx (hunk 5 above)
Comment (kflim) [reply src/components/ChatBox.tsx]: i changed to the outer div having fixed height

Comment (Gabau) [reply src/components/ChatBox.tsx]: Why is fixed height needed anyways? can't we make do with min-height?

File: prisma/postgres/schema.prisma [hunk 6]
```diff
@@ -178,12 +179,13 @@ model QuestionAttempt {
...
 model SessionMessage {
+    id          String  @id @default(cuid())
     sessionId   String
-    senderId    String
-    senderName  String
+    userId    String
+    userName  String
     message     String
     createdAt   DateTime    @default(now())
+    user        User        @relation(fields: [userId], references: [id])
```
Comment (Gabau) [inline prisma/postgres/schema.prisma]: should add a onDelete: Cascade

File: src/components/ChatBox.tsx [hunk 7]
```diff
@@ -20,17 +20,21 @@ const Chatbox = ({
   ] = useSessionComm(sessionId, userId, userName);
 
   return (
-    <div className={className}>
-      <div className="messages-container overflow-y-auto flex flex-col h-[18.5rem]">
-        {allSessionMessages?.map((message, index) => {
+    <div className={className} style={{ height: "22rem" }}>
```
Comment (Gabau) [inline src/components/ChatBox.tsx]: Not really a fan of using fixed height divs. Whats the need for this

**Review by Gabau — COMMENTED** [review]
Also can fix the build? it seems that u nvr push ur change to package json or yarn lock, cause we are missing the react-tabs dependency?

File: src/components/ChatBox.tsx (hunk 7 above)
Comment (kflim) [reply src/components/ChatBox.tsx:23]: To stop the messages container from increasing in height indefinitely until screen height cos of overflow y

File: prisma/postgres/schema.prisma (hunk 6 above)
Comment (kflim) [reply prisma/postgres/schema.prisma]: This one I thought about cos if a user decides to delete their acc in the middle of a session, all the messages on their side would vanish, and actually now I think a better option is on delete set default and then set default user name to be <deleted>

Comment (kflim) [reply prisma/postgres/schema.prisma]: also fixed the build

File: src/components/ChatBox.tsx (hunk 7 above)
Comment (kflim) [reply src/components/ChatBox.tsx]: Refactored height to be 100% on tabs--selected instead of this

File: prisma/postgres/schema.prisma (hunk 6 above)
Comment (Gabau) [reply prisma/postgres/schema.prisma]: Whats wrong with deleting the messages when the user decides to delete their acct tho? 

Also just realised we are refetching every time there is a message, but the whole idea of using a subscription is that we can simply pass the message in the subscription and attach it to the current state

Comment (Gabau) [reply prisma/postgres/schema.prisma]: we should only need to refetch the data on reload, which should be handled by the use query hook

Comment (kflim) [reply prisma/postgres/schema.prisma]: fixed everything mentioned here

File: src/hooks/useSessionComm.ts [hunk 8]
```diff
@@ -25,35 +25,42 @@ export default function useSessionComm(
...
+
+          if (allSessionMessages)
+            allSessionMessages?.push({
+              ...data,
+              id: data.id!,
+              message: data.message,
+              createdAt: data.createdAt!,
+            });
+          // TODO: Fix autoscroll to bottom
+          const messagesContainer =
```
Comment (Gabau) [inline src/hooks/useSessionComm.ts]: Idt this is a responsibilty of the hook, but rather the messages? So the chat messages should default to the bottom right?

Comment (Gabau) [reply src/hooks/useSessionComm.ts]: Like something liddis, https://stackoverflow.com/questions/18614301/keep-overflow-div-scrolled-to-bottom-unless-user-scrolls-up,
if it takes too much time then dun bother tho

File: src/env.mjs [hunk 9]
```diff
@@ -54,7 +54,10 @@ export const env = createEnv({
     MONGO_URL: process.env.MONGO_URL,
     NODE_ENV: process.env.NODE_ENV,
     NEXTAUTH_SECRET: process.env.NEXTAUTH_SECRET,
-    NEXT_PUBLIC_WS_PORT: process.env.NODE_ENV === "production" ? process.env.NEXT_PUBLIC_WS_PORT : "3001",
+    NEXT_PUBLIC_WS_PORT:
```
Comment (Gabau) [inline src/env.mjs]: Is this a formatter change?

**Review by Gabau — APPROVED** [review]
LGTM, not too sure about handling the scroll logic in the hook though.
Alternatively we could return the chat box component from the hook

File: src/hooks/useSessionComm.ts (hunk 8 above)
Comment (kflim) [reply src/hooks/useSessionComm.ts]: yea maybe for another time but i'll remove it from the hook first

File: src/env.mjs (hunk 9 above)
Comment (kflim) [reply src/env.mjs]: this one is automatic by my settings, I can just change it back

**Pull request merged**